In [23]:
# Import the modules
import numpy as np
import pandas as pd
import hvplot.pandas
from pathlib import Path
from sklearn.linear_model import LinearRegression
from sklearn.metrics import balanced_accuracy_score, confusion_matrix, classification_report

---

## Split the Data into Training and Testing Sets

### Step 1: Read the `lending_data.csv` data from the `Resources` folder into a Pandas DataFrame.

In [24]:
# Read the CSV file from the Resources folder into a Pandas DataFrame
df_lending = pd.read_csv(
    Path("./Resources//lending_data.csv")
)
# Review the DataFrame
display(df_lending.head())
display(df_lending.tail())

loan_size  interest_rate  borrower_income  debt_to_income  num_of_accounts  \
0    10700.0          7.672            52800        0.431818                5   
1     8400.0          6.692            43600        0.311927                3   
2     9000.0          6.963            46100        0.349241                3   
3    10700.0          7.664            52700        0.430740                5   
4    10800.0          7.698            53000        0.433962                5   

   derogatory_marks  total_debt  loan_status  
0                 1       22800            0  
1                 0       13600            0  
2                 0       16100            0  
3                 1       22700            0  
4                 1       23000            0

loan_size  interest_rate  borrower_income  debt_to_income  \
77531    19100.0         11.261            86600        0.653580   
77532    17700.0         10.662            80900        0.629172   
77533    17600.0         10.595            80300        0.626401   
77534    16300.0         10.068            75300        0.601594   
77535    15600.0          9.742            72300        0.585062   

       num_of_accounts  derogatory_marks  total_debt  loan_status  
77531               12                 2       56600            1  
77532               11                 2       50900            1  
77533               11                 2       50300            1  
77534               10                 2       45300            1  
77535                9                 2       42300            1

### Step 2: Create the labels set (`y`)  from the “loan_status” column, and then create the features (`X`) DataFrame from the remaining columns.

In [25]:
# Separate the data into labels and features

# Separate the y variable, the labels
y = df_lending['loan_status']

# Separate the X variable, the features
X = df_lending.drop(columns='loan_status')

In [26]:
# Review the y variable Series
print(y.dtype)

int64


In [27]:
# Review the X variable DataFrame
print(X.describe())
print(X.isnull().sum())

          loan_size  interest_rate  borrower_income  debt_to_income  \
count  77536.000000   77536.000000     77536.000000    77536.000000   
mean    9805.562577       7.292333     49221.949804        0.377318   
std     2093.223153       0.889495      8371.635077        0.081519   
min     5000.000000       5.250000     30000.000000        0.000000   
25%     8700.000000       6.825000     44800.000000        0.330357   
50%     9500.000000       7.172000     48100.000000        0.376299   
75%    10400.000000       7.528000     51400.000000        0.416342   
max    23800.000000      13.235000    105200.000000        0.714829   

       num_of_accounts  derogatory_marks    total_debt  
count     77536.000000      77536.000000  77536.000000  
mean          3.826610          0.392308  19221.949804  
std           1.904426          0.582086   8371.635077  
min           0.000000          0.000000      0.000000  
25%           3.000000          0.000000  14800.000000  
50%           4.00

### Step 3: Check the balance of the labels variable (`y`) by using the `value_counts` function.

In [28]:
# Check the balance of our target values
print(y.value_counts())

0    75036
1     2500
Name: loan_status, dtype: int64


### Step 4: Split the data into training and testing datasets by using `train_test_split`.

In [29]:
# Import the train_test_learn module
from sklearn.model_selection import train_test_split

# Split the data using train_test_split
# Assign a random_state of 1 to the function
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=1)
print("Training set size:")
print("X_train: ", X_train.shape)
print("y_train: ", y_train.shape)
print("\nTesting set size:")
print("X_test: ", X_test.shape)
print("y_test: ", y_test.shape)

Training set size:
X_train:  (62028, 7)
y_train:  (62028,)

Testing set size:
X_test:  (15508, 7)
y_test:  (15508,)


---

## Create a Logistic Regression Model with the Original Data

###  Step 1: Fit a logistic regression model by using the training data (`X_train` and `y_train`).

In [30]:
# Import the LogisticRegression module from SKLearn
from sklearn.linear_model import LogisticRegression

# Instantiate the Logistic Regression model
# Assign a random_state parameter of 1 to the model
logistic_model = LogisticRegression(random_state=1)

# Fit the model using training data
logistic_model.fit(X_train, y_train)

LogisticRegression(random_state=1)

### Step 2: Save the predictions on the testing data labels by using the testing feature data (`X_test`) and the fitted model.

In [31]:
# Make a prediction using the testing data
y_pred = logistic_model.predict(X_test)
print(y_pred)

[0 0 0 ... 0 0 0]


### Step 3: Evaluate the model’s performance by doing the following:

* Calculate the accuracy score of the model.

* Generate a confusion matrix.

* Print the classification report.

In [33]:
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report
# Print the balanced_accuracy score of the model
accuracy = accuracy_score(y_test, y_pred)
print("Accuracy Score:", accuracy)


Accuracy Score: 0.9921975754449317


In [34]:
# Generate a confusion matrix for the model
conf_matrix = confusion_matrix(y_test, y_pred)
print("Confusion Matrix:")
print(conf_matrix)

Confusion Matrix:
[[14926    75]
 [   46   461]]


In [35]:
# Print the classification report for the model
class_report = classification_report(y_test, y_pred)
print("Classification Report:")
print(class_report)

Classification Report:
              precision    recall  f1-score   support

           0       1.00      1.00      1.00     15001
           1       0.86      0.91      0.88       507

    accuracy                           0.99     15508
   macro avg       0.93      0.95      0.94     15508
weighted avg       0.99      0.99      0.99     15508



### Step 4: Answer the following question.

**Question:** How well does the logistic regression model predict both the `0` (healthy loan) and `1` (high-risk loan) labels?

**Answer:** The logistic regression model is efficient at identifying healthy loans (0) with high precision, recall, and F1-score are all 100%. 

For high-risk loans (1), the model identifying a portion of actual high-risk loans (91% recall) with a good precision (86%). The F1-score of 0.88 for high-risk loans is commendable, indicating a balanced performance between precision and recall. Logistic regression model predicts both healthy and high-risk loans effectively. 

---

## Predict a Logistic Regression Model with Resampled Training Data

### Step 1: Use the `RandomOverSampler` module from the imbalanced-learn library to resample the data. Be sure to confirm that the labels have an equal number of data points. 

In [36]:
# Import the RandomOverSampler module form imbalanced-learn
from imblearn.over_sampling import RandomOverSampler

# Instantiate the random oversampler model
# # Assign a random_state parameter of 1 to the model
ros = RandomOverSampler(random_state=1)

# Fit the original training data to the random_oversampler model
X_resampled, y_resampled = ros.fit_resample(X_train, y_train)
print(y_resampled.value_counts())

0    60035
1    60035
Name: loan_status, dtype: int64


In [37]:
# Count the distinct values of the resampled labels data
distinct_value_counts = y_resampled.value_counts()
print(distinct_value_counts)

0    60035
1    60035
Name: loan_status, dtype: int64


### Step 2: Use the `LogisticRegression` classifier and the resampled data to fit the model and make predictions.

In [38]:
# Instantiate the Logistic Regression model
# Assign a random_state parameter of 1 to the model
logistic_model_resampled = LogisticRegression(random_state=1)

# Fit the model using the resampled training data
logistic_model_resampled.fit(X_resampled, y_resampled)

# Make a prediction using the testing data
y_pred_resampled = logistic_model_resampled.predict(X_test)
print("Predictions on the testing data labels using resampled data:")
print(y_pred_resampled)

Predictions on the testing data labels using resampled data:
[0 0 0 ... 0 0 0]


### Step 3: Evaluate the model’s performance by doing the following:

* Calculate the accuracy score of the model.

* Generate a confusion matrix.

* Print the classification report.

In [39]:
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report, balanced_accuracy_score
# Print the balanced_accuracy score of the model 
accuracy_resampled = accuracy_score(y_test, y_pred_resampled)
print("Accuracy Score (with resampled data):", accuracy_resampled)

Accuracy Score (with resampled data): 0.9942610265669332


In [40]:
# Generate a confusion matrix for the model
conf_matrix_resampled = confusion_matrix(y_test, y_pred_resampled)
print("Confusion Matrix (with resampled data):")
print(conf_matrix_resampled)

Confusion Matrix (with resampled data):
[[14915    86]
 [    3   504]]


In [41]:
# Print the classification report for the model
class_report_resampled = classification_report(y_test, y_pred_resampled)
print("Classification Report (with resampled data):")
print(class_report_resampled)

Classification Report (with resampled data):
              precision    recall  f1-score   support

           0       1.00      0.99      1.00     15001
           1       0.85      0.99      0.92       507

    accuracy                           0.99     15508
   macro avg       0.93      0.99      0.96     15508
weighted avg       1.00      0.99      0.99     15508



### Step 4: Answer the following question

**Question:** How well does the logistic regression model, fit with oversampled data, predict both the `0` (healthy loan) and `1` (high-risk loan) labels?

**Answer:** logistic regression model that was fitted with the oversampled data suggests that the model performs very well in predicting both 0 (healthy loan) and 1 (high-risk loan) labels. the accuracy was 99%, which is extremely high and the model correctly predicted the loan status in 99% of the cases.  the model seems to have a strong performance, and the resampling seems to have helped the model's ability to predict the minority class (1) effectively. The precision is about 85%,The recall is very high, at about 99%, The F1-score is also high at 92%, 